### PR TITLE
many: support mixed outcomes for permissions in prompting constraints

### PR DESCRIPTION
**Pull request by olivercalder**
This PR is based on #14538, and is tracked internally by https://warthogs.atlassian.net/browse/SNAPDENG-32594. It addresses some of the problems discussed in that PR (such as https://github.com/canonical/snapd/pull/14538#issuecomment-2383774515), and more broadly in https://github.com/canonical/desktop-security-center/issues/74. CC @sminez @juanruitina.

Let rule content constraints have heterogeneous outcomes and lifespans for different permissions in the constraints. As such, convert the list of permissions to a map from permission to permission entry, where the entry holds the outcome, lifespan, and duration/expiration for that particular permission, where previous those were global to the containing rule, rule contents, or patch contents.

However, the existing concept of replying "allow"/"deny" to a particular set of requested permissions is clear and simple. We want to keep outcome, lifespan, and duration as reply-scoped values, not permission-specific, when accepting prompt replies. So we need different types of constraints for prompt replies vs. rule contents.

The motivation behind this is so that we can have only a single rule for any given path pattern. We may have a situation where the user previously replied with "allow read `/path/to/foo`" and they're now prompted for write access, they need to be able to respond with "deny read `/path/to/foo`". If we only support a single outcome for any given rule, then we'd need two rules for the same path `/path/to/foo`. Thus, we need rules to support different outcomes for different permissions.

The same logic applies for lifetimes and expirations, though this adds additional complexity now that the concept of rule expiration is shifted to being permission-specific. We care about expired rules in two primary places: when loading rules from disk, we want to discard any expired rules, and when adding a new rule, we want to discard any expired permission entry for a rule which shares a pattern variant with the new rule. For cases where that expired permission entry had a conflicting outcome, we clearly can't have that, and we want to remove the expired permission entry from its containing rule as well, so as to avoid confusion for the user without them needing to check expiration timestamps. Even if the outcome of the expired entry matches that of the new rule's entry for the same permission, we still want to prune the expired permission from the old rule to avoid confusion. The complexity is around when a notice is recorded for a rule for which some permissions have expired. At the moment, the logic is that a notice is recorded in these cases:

- when a rule is loaded from disk
    - data may be `"removed": "expired"` if all permissions are expired
- when a rule is added
- when a rule is patched
- when a rule is removed (with data `"removed": "removed"`)
- when a rule is found to be expired when attempting to add a new rule

Notably, a notice is not recorded automatically when a permission entry expires. Nor is a notice recorded when a permission is found to be expired, so long as its associated rule still has at least one non-expired permission. Neither pruning an expired permission entry from the rule tree nor from the entry's containing rule results in a notice, even though technically the rule data has changed, since the expired permission has been erased. The rationale is that the semantics of the rule have not changed, since the expiration of that permission was part of the semantics of the rule.

Since durations are used when adding/patching a rule and expirations are used when retrieving a rule, in addition to the differences for prompt replies vs. rule contents, we now need several different variants of constraints:
- `promptConstraints`:
    - path, requested permissions list, available permissions list
    - internal to `requestprompts`, unchanged
- `ReplyConstraints`:
    - path pattern, list of permissions
    - containing `PromptReply` holds outcome/lifespan/expiration
    - unchanged from before, though under a new name
    - converted to a `Constraints` if reply warrants a new rule
- `Constraints`:
    - path pattern, map from permission to outcome, lifespan, duration
    - used when adding rule to the rule DB
    - converted to `RuleConstraints` when the new rule is created
- `RuleConstraints`:
    - path pattern, map from permisison to outcome, lifespan, expiration
    - used when retrieving rules from the rule DB
    - never used when POSTing to the API
- `PatchConstraints`:
    - identical to `Constraints`, but with omitempty fields
    - converted to `RuleConstraints` when the patched rule is created

To support this, we define some new types, including `{,Rule}PermissionMap` and `{,Rule}PermissionEntry`. The latter of these is used in the leaves of the rule DB tree in place of the previous set of rule IDs of rules whose patterns render to a given pattern variant.

Whenever possible, logic surrounding constraints, permissions, and expiration is pushed down to methods on these new types, thus simplifying the logic of their callers.

### CODE REVIEW

**Comment by olivercalder** [comment]
TODO:
- Add unit tests for new constraints types
- Add requestrules unit test for partial rule expiration
- Decide on handling of partial- and fully-expired rules when:
  - Getting rule by ID (currently no expiration checks)
  - Getting all rules (current discards fully expired rules, but does not prune expired permissions)
  - Adding rule which conflicts with expired rule permission (currently prunes expired rule permission and removes rule if all permissions are expired)

**Comment by pedronis** [comment]
@olivercalder this needs a rebase now?

File: interfaces/prompting/requestrules/requestrules.go [hunk 1]
```diff
@@ -41,64 +41,39 @@ import (
...
 //
 // Use the rendered string as the key for this entry, since pattern variants
 // cannot otherwise be easily checked for equality.
 type variantEntry struct {
-	Variant patterns.PatternVariant
-	Outcome prompting.OutcomeType
-	RuleIDs map[prompting.IDType]bool
+	Variant     patterns.PatternVariant
+	Outcome     prompting.OutcomeType
+	RuleEntries map[prompting.IDType]*prompting.RulePermissionEntry
```
Comment (pedronis) [inline interfaces/prompting/requestrules/requestrules.go:76]: why is RulePermissionEntry defined outside of here?

**Review by pedronis — COMMENTED** [review]
did a pass on about half of this, some initial questions/comments

File: interfaces/prompting/requestprompts/requestprompts.go [hunk 2]
```diff
@@ -472,39 +521,43 @@ func (pdb *PromptDB) HandleNewRule(metadata *prompting.Metadata, constraints *pr
...
-
-		// Record all allowed permissions at the time of match, in case a
-		// permission is denied and we need to send back a response.
-		allowedPermission := responseForInterfaceConstraintsOutcome(metadata.Interface, prompt.Constraints, outcome)
-
-		// See if the permission matches any of the prompt's remaining permissions
-		modified := prompt.Constraints.subtractPermissions(constraints.Permissions)
-		if !modified {
-			// No permission was matched
+		// Matched, so at least one permission was satisfied
```
Comment (pedronis) [inline interfaces/prompting/requestprompts/requestprompts.go:534]: I'm not sure I understand the comment vs the code, matched doesn't meant that all outcomes are not deny? it seems the comment needs to be clarified/expanded

File: interfaces/prompting/requestprompts/requestprompts.go [hunk 3]
```diff
@@ -330,21 +374,18 @@ func (pdb *PromptDB) AddOrMerge(metadata *prompting.Metadata, path string, reque
...
-		// Remaining permissions were denied, so allow permissions which were
-		// previously allowed by prompt rules
-		allowedPerms = make([]string, 0, len(constraints.originalPermissions)-len(constraints.remainingPermissions))
+	if len(denied) > 0 {
+		allowedPerms = make([]string, 0, len(constraints.originalPermissions)-len(denied))
 		for _, perm := range constraints.originalPermissions {
-			// Exclude any permissions which were remaining at time of denial
-			if !strutil.ListContains(constraints.remainingPermissions, perm) {
+			if !strutil.ListContains(denied, perm) {
 				allowedPerms = append(allowedPerms, perm)
```
Comment (pedronis) [inline interfaces/prompting/requestprompts/requestprompts.go:389]: no tests in the package reach here apparently

File: interfaces/prompting/constraints.go [hunk 4]
```diff
@@ -30,62 +31,128 @@ import (
...
-// interface, otherwise returns an error.
-func (c *Constraints) ValidateForInterface(iface string) error {
+// ToRuleConstraints validates the receiving Constraints and converts it to
+// RuleConstraints. If the constraints are not valid with respect to the given
+// interface, returns an error.
+func (c *Constraints) ToRuleConstraints(iface string, currTime time.Time) (*RuleConstraints, error) {
+	if c.PathPattern == nil {
+		return nil, prompting_errors.NewInvalidPathPatternError("", "no path pattern")
+	}
+	rulePermissions, err := c.Permissions.ToRulePermissionMap(iface, currTime)
```
Comment (pedronis) [inline interfaces/prompting/constraints.go:52]: ToRulePermissionMap seems to be used only here, I think we want to export only methods that are used more than once/can be used elsewhere

File: interfaces/prompting/constraints.go [hunk 5]
```diff
@@ -108,6 +175,277 @@ func (c *Constraints) ContainPermissions(permissions []string) bool {
...
+
+// RulePermissionEntry holds the outcome associated with a particular permission
+// and the lifespan for which that outcome is applicable.
+//
+// RulePermissionEntry is derived from a PermissionEntry after it has been used
+// along with the rule's timestamp to define the expiration timeouts for any
+// permissions which have a lifespan of "timespan". RulePermissionEntry is what
+// is returned when retrieving rule contents, but PermissionEntry is used when
+// replying to prompts, creating new rules, or modifying existing rules.
+type RulePermissionEntry struct {
```
Comment (pedronis) [inline interfaces/prompting/constraints.go:404]: this seems really the same with PermissionEntry, why the two types?

**Review by pedronis — COMMENTED** [review]
couple more comments. I haven't reviewed requestrules.go changes yet, probably would be good to improve the rest first

File: interfaces/prompting/prompting.go [hunk 6]
```diff
@@ -172,9 +171,6 @@ func (lifespan LifespanType) ValidateExpiration(expiration time.Time, currTime t
 		if expiration.IsZero() {
 			return prompting_errors.NewInvalidExpirationError(expiration, fmt.Sprintf("cannot have unspecified expiration when lifespan is %q", lifespan))
 		}
-		if currTime.After(expiration) {
-			return fmt.Errorf("%w: %q", prompting_errors.ErrRuleExpirationInThePast, expiration)
-		}
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply interfaces/prompting/prompting.go:177]: Now that rules can have mixed outcomes/lifespans/expirations, it may be the case that (e.g. when reading from disk) one permission has expired and another one has not. Rather than throwing an error when an expired permission is seen, it instead is removed from the rule at a later step, and any non-expired permissions remain.

In particular: `ValidateExpiration` is only called once, from within `RulePermissionEntry.validate()`, which is in turn only called once, in `RulePermissionMap.validateForInterface()`. Before calling `entry.validate()`, `validateForInterface` first checks `entry.Expired()`, and if the permission entry has expired, the permission is removed from the permission map at the end of the function, after ensuring that no other errors occurred.

So not quite a simplification, it's a change in the distinction between expired rules and invalid rules now that rules can be "partially expired" but still valid.

File: interfaces/prompting/errors/errors.go [hunk 7]
```diff
@@ -229,3 +231,28 @@ func (e *RuleConflictError) Error() string {
...
+
+// Join returns an error that wraps the given errors.
+// Any nil error values are discarded.
+// Join returns nil if every value in errs is nil.
+//
+// TODO: replace with errors.Join() once we're on golang v1.20+
+//
+// This is a lossy implementation of errors.Join, where only the first non-nil
+// error is preserved in a state which can be unwrapped.
+func Join(errs ...error) error {
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply interfaces/prompting/errors/errors.go:243]: Yes, I believe I have some tests somewhere, I'll migrate them here now that there's a centralized place where the placeholder `errors.Join` is defined. Thanks!

File: interfaces/prompting/requestrules/requestrules.go (hunk 1 above)
Comment (olivercalder) [reply interfaces/prompting/requestrules/requestrules.go:76]: A `RulePermissionMap` is a map from permission (string) to `RulePermissionEntry`, so to avoid a circular dependency, `RulePermissionEntry` must be defined in the same place `RulePermissionMap` is (or in a lower level). But at the same time, `RuleConstraints` contains a `RulePermissionMap`, so `RulePermissionMap` must be defined in the same place or lower in the chain than `RuleConstraints`.

An argument could be made for defining `RulePermissionMap` and `RuleConstraints` in `requestrules` instead of `interfaces/prompting`. But `requestprompts` uses `RuleConstraints` as well, so then `requestprompts` would need to import `requestrules`, which is fine but a bit weird to me.

More importantly though, `Constraints` are validated by being converted into `RuleConstraints`. So wherever `Constraints` is defined needs to import whatever defined `RuleConstraints`. So I think `RuleConstraints`, `RulePermissionMap`, and `RulePermissionEntry` all need to be defined in `interfaces/prompting`.

An alternative could be to have separate validation and conversion functions, but these would be essentially identical (and the latter would need to call the former again, probably), so it seems to me like the code overlap would be great, and the conversion functions defined in `requestrules` would need to operate on the internals of those structs as defined in `interfaces/prompting`. So I think it's best to have all the similarly structured and related types defined in one place.

File: interfaces/prompting/constraints.go (hunk 4 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go:52]: Yes, this method can be unexported, thanks! I'll check if I can find any more which should not be exported.

File: interfaces/prompting/constraints.go (hunk 5 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go:404]: `RulePermissionEntry` has an `Expiration` field while `PermissionEntry` has a `Duration` field. When a user replies to a prompt or tries to create or patch a rule directly over the rules API, they always act in terms of duration, rather than expiration. Internally, those durations are validated and then converted to an expiration. But the client should never specify an expiration timestamp when POSTing or PATCHing to the API.

File: interfaces/prompting/requestprompts/requestprompts.go (hunk 2 above)
Comment (olivercalder) [reply interfaces/prompting/requestprompts/requestprompts.go:534]: Matched means the path pattern of the rule matched the path of the prompt, and at least one permission from the rule matched at least one permission from the prompt. Whether all or just some of the permissions were matched, and whether each was allowed or denied, is what the other return values of `matched, satisfied, denied, err := prompt.Constraints.applyRuleConstraints(constraints)` indicates.

But I agree this is rather confusing. `matched`, `satisfied`, and `denied` are not the same types, and there is some implication between each.

I think the complexity comes down to the interaction between the way `buildResponse`, `applyRuleConstraints`, and their callers interact. A lot of it is because of the old way these used to work, which is no longer the case, so they can be simplified. I'll work on that.

**Comment by olivercalder** [comment]
We need to be careful to support both the old and new rule formats during a transition period. The internals should use the new system, but we'll need to map the old format to the new structure, and provide a means of working in the old format over the API.

**Comment by olivercalder** [comment]
I'd definitely appreciate some feedback from the Desktop team about how best to handle backwards compatibility for the client :)

CC @sminez @d-loose

**Comment by olivercalder** [comment]
As discussed with Desktop, the plan is for the `snapd.dart` to replace the explicit constraints format (currently specific to the home interface) with a generic constraints json blob. Then it's up to the client to parse the json blob and handle it accordingly.

Regarding backwards compatibility, since old-format rules can be converted to new-format rules, the plan is for clients of snapd to map from old-format to new-format and pass that on to downstream clients, so they can immediately begin working with the new format. I believe `snapd.dart` will be responsible for doing this mapping before passing the rules on to the security center, but I want to check with @sminez on this first.

As for compatibility on the snapd side, since there are currently no clients which send rule contents directly to snapd (e.g. for creating/patching rules directly, as opposed to implicitly creating rules via prompt replies), we aren't concerned with snapd accepting old-format rule contents over the API. Any client which wishes to POST rule contents to snapd should use the new format, as we expect this PR to be merged and snapd to be using the new format well before any clients will wish to do so.

The spec for the changes to the rule format can be found here: https://docs.google.com/document/d/19m27d-YK9TdGCucZyjDzHYsQqICE2RE2yPia7sTg7Ls/edit

File: interfaces/prompting/constraints.go [hunk 8]
```diff
@@ -97,9 +164,9 @@ func (c *Constraints) Match(path string) (bool, error) {
 	return match, nil
 }
 
-// ContainPermissions returns true if the constraints include every one of the
-// given permissions.
-func (c *Constraints) ContainPermissions(permissions []string) bool {
+// ContainPermissions returns true if the permission list in the constraints
+// includes every one of the given permissions.
+func (c *ReplyConstraints) ContainPermissions(permissions []string) bool {
```
Comment (pedronis) [inline interfaces/prompting/constraints.go:169]: iiuc ReplyContraints is not really validated unless via ToConstraints which makes me think that we really want ReplyContraints to not have any methods except the conversion and move the methods even if annoying to Constraints. That should force changing some of the code I'm right now slightly uncomfortable with.

**Review by pedronis — COMMENTED** [review]
a comment and also this has conflicts now

File: interfaces/prompting/constraints.go (hunk 8 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go:169]: I think that makes sense: `ReplyConstraints.ContainPermissions` and `ReplyConstraints.Match` can be moved to methods on `Constraints` instead.

The rationale for having those methods on `ReplyConstraints` instead of `Constraints` is that one should never call such methods on any constraints except ones which are received as part of a reply. But we can add a comment to that effect instead.

Comment (olivercalder) [reply interfaces/prompting/constraints.go:169]: You were also right about moving `responseForInterfaceConstraintsOutcome` in #14390, it's making the rebase much more difficult...

**Comment by olivercalder** [comment]
Need this too:
```
Dec 06 13:28:33 autopkgtest kernel: audit: type=1400 audit(1733513313.748:155): apparmor="DENIED" operation="create" class="net" profile="snap.test-snapd-audit-control.audit-rate" pid=15715 comm="auditctl" family="netlink" sock_type="raw" protocol=9 requested="create" denied="create"
```

**Comment by pedronis** [comment]
On Fri, 6 Dec 2024 at 20:30, Oliver Calder ***@***.***> wrote:

> Need this too:
>
> Dec 06 13:28:33 autopkgtest kernel: audit: type=1400 audit(1733513313.748:155): apparmor="DENIED" operation="create" class="net" profile="snap.test-snapd-audit-control.audit-rate" pid=15715 comm="auditctl" family="netlink" sock_type="raw" protocol=9 requested="create" denied="create"
>
>
Wrong PR for this comment?

File: interfaces/prompting/requestprompts/requestprompts.go [hunk 9]
```diff
@@ -100,30 +100,36 @@ func (p *Prompt) MarshalJSON() ([]byte, error) {
...
-		allow = false
-		logger.Noticef("internal error: failed to compute prompting outcome: %v", err)
-	}
+// buildResponse creates a listener response based on the given interface,
+// prompt constraints, and list of denied permissions.
+//
+// The response is the AppArmor permission which should be allowed. This
+// corresponds to the originally requested permissions from the prompt
+// constraints, except with all denied permissions removed.
+func buildResponse(iface string, constraints *promptConstraints, denied []string) any {
```
Comment (pedronis) [inline interfaces/prompting/requestprompts/requestprompts.go:127]: given the usage it seems this should be a method on promptConstraints

File: interfaces/prompting/requestprompts/requestprompts.go [hunk 10]
```diff
@@ -195,20 +201,65 @@ func (pc *promptConstraints) equals(other *promptConstraints) bool {
...
+// Returns whether the prompt constraints were modified, whether the prompt
+// requires a response (either because all permissions were allowed or at least
+// one permission was denied), and the list of any permissions which were
+// denied. If an error occurs, it is returned, and the other parameters can be
+// ignored.
+//
+// If the path pattern does not match the prompt path, or the rule constraints
+// do not include any of the remaining prompt permissions, then modified is
+// false, and no changes are made to the prompt constraints.
+func (pc *promptConstraints) applyRuleConstraints(constraints *prompting.RuleConstraints) (modified, respond bool, denied []string, err error) {
```
Comment (pedronis) [inline interfaces/prompting/requestprompts/requestprompts.go:216]: maybe coveredByRule is a better concept than modified here

File: interfaces/prompting/requestrules/requestrules.go [hunk 11]
```diff
@@ -1038,16 +1000,28 @@ func (rdb *RuleDB) RemoveRulesForSnapInterface(user uint32, snap string, iface s
...
-// Any of the parameters which are equal to the default/unset value for their
-// types are left unchanged from the existing rule. Even if the given new rule
-// contents exactly match the existing rule contents, the timestamp of the rule
-// is updated to the current time. If there is any error while modifying the
-// rule, the rule is rolled back to its previous unmodified state, leaving the
-// database unchanged. If the database is changed, it is saved to disk.
-func (rdb *RuleDB) PatchRule(user uint32, id prompting.IDType, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (r *Rule, err error) {
+// Permission entries must be provided as complete units, containing both
+// outcome and lifespan (and duration, if lifespan is timespan).
+// XXX: does API unmarshalling ensures this, or do we need explicit checks?
```
Comment (pedronis) [inline interfaces/prompting/requestrules/requestrules.go:1013]: what's this XXX about exactly?

File: interfaces/prompting/requestrules/requestrules.go [hunk 12]
```diff
@@ -507,37 +490,41 @@ func (rdb *RuleDB) addRulePermissionToTree(rule *Rule, permission string) []prom
...
+	expiredData := map[string]string{"removed": "expired"}
+	for ruleID := range partiallyExpiredRules {
+		maybeExpired, err := rdb.lookupRuleByIDForUser(rule.User, ruleID)
+		if err != nil {
+			// Error shouldn't occur. If it does, the rule was already removed
+			continue
+		}
+		// Already removed the rule's permission from the tree, let's remove
+		// it from the rule as well
+		delete(maybeExpired.Constraints.Permissions, permission)
```
Comment (pedronis) [inline interfaces/prompting/requestrules/requestrules.go:502]: is it relevant to do this delete if we hit the code below that removes the entire rule? shouldn't we do this only in the non-expired case?

**Review by pedronis — COMMENTED** [review]
did another pass, the partiallyExpiredRules code is a bit hard to follow tbh

File: interfaces/prompting/requestprompts/requestprompts.go (hunk 10 above)
Comment (olivercalder) [reply interfaces/prompting/requestprompts/requestprompts.go:216]: To me, "covered" seems to imply the prompt constraints were entirely satisfied by the rule constraints, but what `modified` indicates here is that at least one permission was satisfied. `respond` indicates that the prompt constraints were satisfied by the rule constraints, either because at least one permission was denied or all permissions were allowed.

How about `affectedByRule` instead of `coveredByRule`, to replace `modified`?

File: interfaces/prompting/requestrules/requestrules.go (hunk 11 above)
Comment (olivercalder) [reply interfaces/prompting/requestrules/requestrules.go:1013]: Let's say the existing rule has the following constraints:
```json
"constraints": {
    "path-pattern": "/foo/bar",
    "permissions": {
        "read": {
            "outcome": "allow",
            "lifespan": "forever"
        },
        "write": {
            ...
        }
    }
}
```
and they want to switch the outcome of the `"read"` permission from `"allow"` to `"deny"`. We need them to include the complete permission entry for "read" in the patch constraints (though they can omit other unchanged fields, such as `"path-pattern"`, and the `"write"` permission):
```json
"constraints": {
    "permissions": {
        "read": {
            "outcome": "deny",
            "lifespan": "forever" // can't omit this
        }
    }
}
```

The resulting constraints are then:
```json
"constraints": {
    "path-pattern": "/foo/bar",
    "permissions": {
        "read": {
            "outcome": "deny",
            "lifespan": "forever"
        },
        "write": {
            ...
        }
    }
}
```

Since the only `omitempty` field in the permission entry is `"duration"`, if there's a permission entry for `"read"` in the patch constraints, the unmarshaller will enforce that it contains both the `"outcome"` and `"lifespan"` fields --- one can't omit the `"lifespan"` field and assume that the existing lifespan will be reused.

So this XXX can be removed. I'll add a comment summarizing the explanation here.

File: interfaces/prompting/requestrules/requestrules.go (hunk 12 above)
Comment (olivercalder) [reply interfaces/prompting/requestrules/requestrules.go]: The end result is the same (since pruning an expired permission from a rule which is not fully expired does not result in a notice), but might as well only do this if `!maybeExpired.expired()`. I've made the change.

File: interfaces/prompting/requestprompts/requestprompts.go (hunk 10 above)
Comment (pedronis) [reply interfaces/prompting/requestprompts/requestprompts.go:191]: affectedByRule is also ok

File: interfaces/prompting/constraints.go [hunk 13]
```diff
@@ -326,13 +330,13 @@ func (pm RulePermissionMap) validateForInterface(iface string, currTime time.Tim
 			invalidPerms = append(invalidPerms, perm)
 			continue
 		}
+		if err := entry.validate(); err != nil {
+			errs = append(errs, err)
```
Comment (pedronis) [inline interfaces/prompting/constraints.go:334]: should we "continue" here?

**Review by pedronis — COMMENTED** [review]
one question

File: interfaces/prompting/constraints.go [hunk 14]
```diff
@@ -97,17 +129,333 @@ func (c *Constraints) Match(path string) (bool, error) {
...
+}
+
+// PatchConstraints hold information about the applicability of the modified
+// rule to particular paths or permissions. A request will be matched by the
+// constraints if the requested path is matched by the path pattern (according
+// to bash's globstar matching) and one or more requested permissions are
+// denied in the permission map, or all of the requested permissions are
+// allowed in the map. When snapd modifies the rule, it converts
+// PatchConstraints to RuleConstraints, using the rule's existing constraints
+// wherever a field is omitted.
```
Comment (bboozzoo) [inline interfaces/prompting/constraints.go:191]: feels like a shorter comment would be less confusing
```suggestion
// PatchConstraints hold information about the applicability of the modified
// rule to particular paths or permissions. When snapd modifies the rule, 
// it converts PatchConstraints to RuleConstraints, using the rule's existing
// constraints wherever a field is omitted.
```

File: interfaces/prompting/constraints.go [hunk 15]
```diff
@@ -97,17 +129,333 @@ func (c *Constraints) Match(path string) (bool, error) {
...
+//
+// Any permissions which are omitted from the new permission map are left
+// unchanged from the existing rule. To remove an existing permission from the
+// rule, the permission should map to null.
+type PatchConstraints struct {
```
Comment (bboozzoo) [inline interfaces/prompting/constraints.go:196]: the name feels slightly confusing, as if the structure carries constraints for a patch, and not a patch for constraints, how about `ConstraintsPatch` or `RuleConstraintsPatch` then?

File: interfaces/prompting/constraints.go [hunk 16]
```diff
@@ -30,62 +31,93 @@ import (
...
-	return nil
+	return ruleConstraints, nil
+}
+
+// RuleConstraints hold information about the applicability of an existing rule
+// to particular paths or permissions. A request will be matched by the rule
+// constraints if the requested path is matched by the path pattern (according
+// to bash's globstar matching) and one or more requested permissions are denied
+// in the permission map, or all of the requested permissions are allowed in the
+// map.
```
Comment (bboozzoo) [inline interfaces/prompting/constraints.go:100]: not sure it's useful to repeat the same thing in every doc comment.
```suggestion
// to particular paths or permissions.
```

File: interfaces/prompting/constraints.go [hunk 17]
```diff
@@ -30,62 +31,93 @@ import (
...
+	ruleConstraints := &RuleConstraints{
+		PathPattern: c.PathPattern,
+		Permissions: rulePermissions,
 	}
-	c.Permissions = newPermissions
-	return nil
+	return ruleConstraints, nil
+}
+
+// RuleConstraints hold information about the applicability of an existing rule
```
Comment (bboozzoo) [inline interfaces/prompting/constraints.go:95]: feels like this should be name ConstraintsRule?

File: interfaces/prompting/constraints.go (hunk 13 above)
Comment (bboozzoo) [reply interfaces/prompting/constraints.go:334]: I think so too

File: interfaces/prompting/constraints.go (hunk 17 above)
Comment (pedronis) [reply interfaces/prompting/constraints.go:95]: no, that doesn't sound right either.  Constraints that go into a rule seems fine to be called RuleConstraints, I understand that it reads also as Constraints for a Rule which might mean a couple of things, but I think contextually is clear enough

File: interfaces/prompting/errors/errors.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply interfaces/prompting/errors/errors.go:243]: can't we put it in strutil for now?

File: interfaces/prompting/constraints.go (hunk 13 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go:334]: Good catch, thanks. I didn't fix the `continue`s when I reordered the validation/expiration checks. I'll confirm that the other place things were reordered is correct as well.

Comment (olivercalder) [reply interfaces/prompting/constraints.go:334]: Other one is correct. I've fixed this case, will push soon.

File: interfaces/prompting/constraints.go [hunk 18]
```diff
@@ -97,17 +129,333 @@ func (c *Constraints) Match(path string) (bool, error) {
...
+	if !ok {
+		return nil, prompting_errors.NewInvalidInterfaceError(iface, availableInterfaces())
+	}
+	if len(c.Permissions) == 0 {
+		return nil, prompting_errors.NewPermissionsListEmptyError(iface, availablePerms)
+	}
+	var invalidPerms []string
+	permissionMap := make(PermissionMap, len(c.Permissions))
+	for _, perm := range c.Permissions {
+		if !strutil.ListContains(availablePerms, perm) {
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply interfaces/prompting/constraints.go:164]: Permissions are not sorted lexicographically, they're ordered according to the order in which we'd like to present them to the user: `["read", "write", "execute"]`.

File: interfaces/prompting/constraints.go [hunk 19]
```diff
@@ -97,17 +129,333 @@ func (c *Constraints) Match(path string) (bool, error) {
...
+// To remove an an existing permission from the rule, the permission should map
+// to null in the permission map of the patch constraints.
+//
+// The existing rule constraints should never be modified.
+func (c *PatchConstraints) PatchRuleConstraints(existing *RuleConstraints, iface string, currTime time.Time) (*RuleConstraints, error) {
+	ruleConstraints := &RuleConstraints{
+		PathPattern: c.PathPattern,
+	}
+	if c.PathPattern == nil {
+		ruleConstraints.PathPattern = existing.PathPattern
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply interfaces/prompting/constraints.go:217]: Since we never modify the data in the `PathPattern` struct after it's been constructed, and this is simply a pointer to that heap-allocated struct, it should be fine to do this I think.

In particular, the path pattern was allocated independently of any constraints, so we won't prevent the old constraints from being garbage collected, if that's the concern.

File: interfaces/prompting/constraints.go [hunk 20]
```diff
@@ -97,17 +129,333 @@ func (c *Constraints) Match(path string) (bool, error) {
...
-// given permissions.
-func (c *Constraints) ContainPermissions(permissions []string) bool {
-	for _, perm := range permissions {
-		if !strutil.ListContains(c.Permissions, perm) {
+// ReplyConstraints hold information about the applicability of a reply to
+// particular paths or permissions. Upon receiving the reply, snapd converts
+// ReplyConstraints to Constraints.
+type ReplyConstraints struct {
+	PathPattern *patterns.PathPattern `json:"path-pattern"`
+	Permissions []string              `json:"permissions"`
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply interfaces/prompting/constraints.go:137]: We don't check this, and we don't particularly care if the client response with something like `allow ["read", "read"]`. `ReplyConstraints` are validated during `ToConstraints`, where we iterate through the permissions in the `Permissions` list and add a permission entry to the permission map for each permission. Duplicate permissions in the list means we'll re-add the same `key:value` to the map twice, but the key and value are identical in this case, so that's no problem.

File: interfaces/prompting/constraints.go (hunk 16 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go:100]: Agreed, though I think I'd rather have the explanation here, as `RuleConstraints` are what is used to actually match against prompts. `Constraints` are the more ergonomic (using `Duration` instead of `Expiration`) form of `RuleConstraints`, and they're converted to the latter before being matched against any requests. I'll remove the verbose explanation from `Constraints` and note that `Constraints` are converted into `RuleConstraints`.

File: interfaces/prompting/constraints.go (hunk 15 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go:189]: What are your thoughts on this @pedronis ?

File: interfaces/prompting/errors/errors.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (olivercalder) [reply interfaces/prompting/errors/errors.go:243]: I'll move it to strutil for now. Hopefully we can move to go 1.20+ soon though...

**Review by pedronis — APPROVED** [review]
thank you

**Review by bboozzoo — APPROVED** [review]
LGTM

File: interfaces/prompting/constraints.go [hunk 21]
```diff
@@ -30,62 +31,90 @@ import (
...
-// is matched by the path pattern (according to bash's globstar matching) and
-// the requested permissions are contained in the constraints' permissions.
+// Constraints hold information about the applicability of a new rule to
+// particular paths and permissions. When creating a new rule, snapd converts
+// Constraints to RuleConstraints.
 type Constraints struct {
-	PathPattern *patterns.PathPattern `json:"path-pattern,omitempty"`
-	Permissions []string              `json:"permissions,omitempty"`
+	PathPattern *patterns.PathPattern `json:"path-pattern"`
+	Permissions PermissionMap         `json:"permissions"`
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply interfaces/prompting/constraints.go:39]: Yes. The `desktop-security-center` is the only supported client at the moment, and it will accept either the old or the new format and always map the result to the new format, since the new format can express a proper superset of what the old format can express. Because the feature is experimental and the API is not yet officially documented, we feel comfortable making this breaking change, which the only "official" client already handles correctly.

**Comment by olivercalder** [comment]
Test failures:
- openstack:debian-12-64:tests/completion/ -- dpkg problem, unrelated
- openstack:opensuse -- openstack problems, unrelated
- google:ubuntu-16.04-64:tests/main/degraded -- unrelated
- google:ubuntu-20.04-64:tests/unit/go:gcc -- unrelated, fix in the works by Katie
- google-arm:ubuntu-core-22-arm-64:tests/core/snapd-refresh-undo -- unrelated, fix in the works by Katie
